tests/test_type/test_array.py

```diff
@@ -41,6 +41,24 @@ def test_array():
     assert not isinstance(m.Out(m.Array[2, m.Bit])(), m.In(m.Array[2, m.Bit]))
 
 
+def test_qualify_array():
+    assert str(m.In(Array)) == "In(Array)"
+    assert str(m.Out(m.In(Array))) == "Out(Array)"
+    assert str(m.In(Array)[5, m.Bit]) == "Array[5, In(Bit)]"
+    assert str(m.Out(m.In(Array))[5, m.Bit]) == "Array[5, Out(Bit)]"
+
+    # Array qualifer overrides child qualifer
+    assert str(m.In(Array)[5, m.Out(m.Bit)]) == "Array[5, In(Bit)]"
```

Comment:
if we have these weird nested direction qualifiers, I feel like we should log/warn what rules we're applying. (Don't disagree with the rules, just want them to be transparent to the user who maybe isn't aware of this complexity.)